SQL: Extend the optimisations for equalities

bpintea opened the pull request:
This PR supplements the optimisations of equalities in conjunctions
and disjunctions:
* for conjunctions, the existing optimizations with ranges are extended
with not-equalities and inequalities; these lead to a fast resolution,
the conjunction either being evaluate to a FALSE, or the non-equality
conditions being dropped as superfluous;
* optimisations for disjunctions are added to be applied against ranges,
inequalities and not-equalities; these lead to disjunction either
becoming TRUE or the equality being dropped, either as superfluous or
merged into a range/inequality.

Adresses #49637

bpintea committed:
Extend the optimizations for equalities

This commit supplements the optimisations of equalities in conjunctions
and disjunctions:
* for conjunctions, the existing optimizations with ranges are extended
with not-equalities and inequalities; these lead to a fast resolution,
the conjunction either being evaluate to a FALSE, or the non-equality
conditions being dropped as superfluous;
* optimisations for disjunctions are added to be applied against ranges,
inequalities and not-equalities; these lead to disjunction either
becoming TRUE or the equality being dropped, either as superfluous or
merged into a range/inequality.

elasticmachine commented:
Pinging @elastic/es-search (:Search/SQL)

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+        // a = 2 OR a != 5 -> TRUE; a = 2 OR a = 5 -> nop
+        private Expression propagate(Or or) {
+            List<Expression> exps = new ArrayList<>();
+            List<Equals> frEquals = new ArrayList<>(); // foldable right term Equals
```
I see that fr comes from foldable right - I would remove the prefix since there's no foldable left and the collections are used for compacting or selecting the expressions.

bpintea replied:
Sure, removed it. I didn't like it either, tbh, but was trying to avoid the confusion as to why not all references of a certain type are part of the respective container.  But with the added comment (and code) it should be clear enough, indeed.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+            List<BinaryComparison> frInequalities = new ArrayList<>(); // foldable right term (=limit) BinaryComparision
+
+            // split expressions by type
+            for (Expression ex : Predicates.splitOr(or)) {
```
Isn't this handled already by the folding rule  + BooleanSimplification?

bpintea replied:
I believe it's not, no.
(But I'll happily be corrected.)

costin replied:
Or at:
https://github.com/elastic/elasticsearch/blob/54c1a59650833361edb85d3a7e3bdb933d9571ba/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L838
And at: 
https://github.com/elastic/elasticsearch/blob/54c1a59650833361edb85d3a7e3bdb933d9571ba/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L809

That is each individual and/or is evaluated if possible

bpintea replied:
I had a look at that, but the simplification added by the code isn't covered there. Maybe that's what the [`// TODO: eliminate conjunction/disjunction`](https://github.com/elastic/elasticsearch/blob/54c1a59650833361edb85d3a7e3bdb933d9571ba/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L876) was hinting to?
 
(BTW and related to the last suggestions below, the comments [`(a || b) && (a || c) => a && (b || c)`](https://github.com/elastic/elasticsearch/blob/54c1a59650833361edb85d3a7e3bdb933d9571ba/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L817) and [`(a && b) || (a && c) => a || (b & c)`](https://github.com/elastic/elasticsearch/blob/54c1a59650833361edb85d3a7e3bdb933d9571ba/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L855) will also need fixing (the code does the right thing, though) -- I'll address this too.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+                        updated = true;
+                    } else if (lowerComp != null && upperComp != null) {
+                        if (0 < lowerComp && upperComp < 0) { // a = 2 OR 1 < a < 3
+                            iterEq.remove(); // equality is superfluous
```
It's worth reviewing the code that does removal through position (`for (int i = ...) { ranges.remove(i); } `) since it's incorrect as after a removal it skips the next item as the counter is not decremented.
Moving that to use iterator removal is better (though more verbose).

bpintea replied:
Right, the `ranges` index-based looping doesn't do removals in the `ranges` lists, only updates at current index (for cases like `a = 2 OR 2 < a < ? -> 2 <= a < ?`); in this case, the removal is operated on the `equals` list, but that's done through the iterator.
(But let me know please if I still miss something.)

costin replied:
The removal since safe since it is followed by an add the iteration is not affected but inside `CombineBinaryComparisons` there are some removals that are not compensated for 
https://github.com/elastic/elasticsearch/blob/54c1a59650833361edb85d3a7e3bdb933d9571ba/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L1138
which means the next entry in the `bcs` list might skipped.
Worth reviewing (and documenting).

bpintea replied:
Ah, I see. Sure, I'll address that in a separate issue.

costin reviewed (approved):
Left some comments but otherwise LGTM!

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+                // Equals OR Inequality
+                for (int i = 0; i < frInequalities.size(); i ++) {
+                    BinaryComparison bc = frInequalities.get(i);
+                    if (! eq.left().semanticEquals(bc.left())) {
```
`== false`

bpintea replied:
Updated.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+                            continue;
+                        } else if (comp == 0 && bc instanceof GreaterThan) { // a = 2 OR a > 2 -> a >= 2
+                                frInequalities.set(i, new GreaterThanOrEqual(bc.source(), bc.left(), bc.right()));
+                        } // else (0 < comp) : // a = 3 OR a > 2 -> a > 2
```
The commented potential condition here - `0 < comp` - is not entirely true. A potential `else` branch can also happen for `bc instanceof GreaterThanOrEqual`. I don't think it's changing the logic, but just saying.

bpintea replied:
Right, better be thorough also in comments. Fixed in both affected places.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+                        iterEq.remove(); // update range with equality instead or simply superfluous
+                        updated = true;
+                    } else if (bc instanceof LessThan || bc instanceof LessThanOrEqual) {
+                        if (0 < comp) { // a = 2 OR a < 1 -> nop
```
Can you, please, be consistent in the code format? The previous `if` branch used `(comp < 0)` and then `else if`. In this one you are switching places between variable and constant and there is no `else`.

bpintea replied:
I generally prefer the "0x axis" spacial ordering for faster understanding - the same reason why ranges are mostly given like `a < X < b` and less so the other way around -, but I've updated the ordering here.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
             }
 
-            return changed ? Predicates.combineAnd(CollectionUtils.combine(exps, equals, ranges)) : and;
+            return updated ? Predicates.combineOr(CollectionUtils.combine(exps, frEquals, frNotEquals, frInequalities, ranges)) : or;
```
`updated == true ? Predicates.....`

astefan reviewed (changes requested):
Nice simplification work.
I have mostly cosmetic related comments, but also one that I believe needs a second look over.

bpintea committed:
Adress review notes

* Fix the bug around wrongly optimizing 'a=2 OR a!=?', which only yields
TRUE for same values in equality and inequality.
* Var renamings, code style adjustments, comments corrections.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -976,14 +979,20 @@ private Expression literalToTheRight(BinaryOperator<?, ?, ?, ?> be) {
-            List<BinaryComparison> equals = new ArrayList<>();
+            // Only equalities, not-equalities and inequalities with a foldable .right are extracted separately;
+            // the others go into the general 'exps'.
+            List<BinaryComparison> frEquals = new ArrayList<>();
```
Right, it wasn't a great solution.. I've gone ahead and adopted Costin's solution to simply remove the prefix. I've considered the `foldableRight` prefix, but thought that having it would slightly reduce the var distinguishability due to lenght (i.e. `foldableRightEqual` vs `foldableRightNotEqual` vs `foldableRightInequalities` etc.); and `fold` isn't super clear imo: foldable vs folded vs..?
But no strong feelings about it.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+                // evaluate all inequalities against the Equal
+                for (Iterator<BinaryComparison> iter = frInequalities.iterator(); iter.hasNext(); ) {
+                    BinaryComparison bc = iter.next();
+                    if (! eq.left().semanticEquals(bc.left())) {
```
> I think we prefer the eq.left().semanticEquals(bc.left()) == false format for better visibility.

Ah, correct, I think it's even in a guideline somewhere the format with just a variable.

>  I would have kept the same "style" as the previous conditional statements and not use `continue`

I find the more convoluted loops easier to read when flattened, but it's only a slight preference, so I've refactored it in this case.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+                for (Iterator<NotEquals> iter = frNotEquals.iterator(); iter.hasNext(); ) {
+                    NotEquals neq = iter.next();
+                    if (eq.left().semanticEquals(neq.left())) {
+                        Integer comp = BinaryComparison.compare(eq.right().fold(), neq.right().fold());
```
Right! Thanks.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+                }
+            }
+
+            return changed ? Predicates.combineAnd(CollectionUtils.combine(exps, frEquals, frNotEquals, frInequalities, ranges)) : and;
```
I'd keep this as is, inline with the style of the rest of the class (my change only affected the parameters passed to `CollectionUtils.combine`).

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+
+                // Equals OR NotEquals
+                for (NotEquals neq : frNotEquals) {
+                    if (eq.left().semanticEquals(neq.left())) { // a = 2 OR a != ? -> TRUE
```
Good spot, thanks!
The `TRUE` is only valid if the disjunction terms refer the same value. I've fixed it.

costin replied:
Please add a test to double check this is the case.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+                // Equals OR Range
+                for (int i = 0; i < ranges.size(); i ++) { // might modify list, so use index loop
+                    Range range = ranges.get(i);
+                    if (! eq.left().semanticEquals(range.value())) {
```
Updated.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
+        // a = 2 OR a > 3 -> nop; a = 2 OR a > 1 -> a > 1
+        // a = 2 OR a < 3 -> a < 3; a = 2 OR a < 1 -> nop
+        // a = 2 OR 3 < a < 5 -> nop; a = 2 OR 1 < a < 3 -> 1 < a < 3; a = 2 OR 0 < a < 1 -> nop
+        // a = 2 OR a != 5 -> TRUE; a = 2 OR a = 5 -> nop
```
Indeed, thanks. Will update along the `a = 2 OR a != 5 -> a != 5` change.

costin commented:
As a separate improvements, we should move away from returning `Literal.TRUE/FALSE` since these constants don't preserve the `Source`, which typically is not needed but why drop it.
Instead the `Source` of the folded expression should be kept around, e.g.:
```java
return FALSE -> return Literal.of(bc, Boolean.FALSE);
```

astefan reviewed (approved):
LGTM

bpintea committed:
Address further review comments. Extend optim.

- fix a few code comments;
- extend the Equals OR NotEquals optimitsation (a=2 OR a!=5 -> a!=5);
- extend the Equals OR Range optimisation on limits equality (a=2 OR
  2<=a<5 -> 2<=a<5);
- in case an equality is being removed in a conjunction, the rest of
  possible optimisations to test is now skipped.

bpintea committed:
rename one var for better legiblity

- s/rmEqual/removeEquals

bpintea committed:
Merge remote-tracking branch 'upstream/master' into enh/propagate-equals-optimisations

costin commented:
> As a separate improvements, we should move away from returning `Literal.TRUE/FALSE` since these constants don't preserve the `Source`, which typically is not needed but why drop it.
> Instead the `Source` of the folded expression should be kept around, e.g.:
> 
> ```java
> return FALSE -> return Literal.of(bc, Boolean.FALSE);
> ```

I've addressed that in #50962
